File: src/Ycs/Types/AbstractType.cs
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using Ycs.Structs;
using Ycs.Contracts;
using Ycs.Utils;

namespace Ycs.Types
{
    public class YEventArgs
    {
        public YEventArgs(YEvent evt, ITransaction transaction)
        {
            Event = evt;
            Transaction = transaction;
        }

        public YEvent Event { get; }
        public ITransaction Transaction { get; }
    }

    public class YDeepEventArgs
    {
        public YDeepEventArgs(IList<YEvent> events, ITransaction transaction)
        {
            Events = events;
            Transaction = transaction;
        }

        public IList<YEvent> Events { get; }
        public ITransaction Transaction { get; }
    }

    public class AbstractType
    {
        public Item Item { get; set; }
        public Item Start { get; set; }
        public IDictionary<string, Item> Map { get; set; } = new Dictionary<string, Item>();

        public event EventHandler<YEventArgs> EventHandler;
        public event EventHandler<YDeepEventArgs> DeepEventHandler;

        public YDoc Doc { get; protected set; }
        public AbstractType Parent => Item != null ? Item.Parent as AbstractType : null;

        public virtual int Length { get; set; }

        public virtual void Integrate(YDoc doc, Item item)
        {
            Doc = doc;
            Item = item;
        }
        public virtual AbstractType InternalCopy() { throw new NotImplementedException(); }
        public virtual AbstractType InternalClone() { throw new NotImplementedException(); }

        public virtual void Write(IUpdateEncoder encoder) { throw new NotImplementedException(); }

        /// <summary>
        /// Call event listeners with an event. This will also add an event to all parents
        /// for observeDeep handlers.
        /// </summary>
        public virtual void CallTypeObservers(ITransaction transaction, YEvent evt)
        {
            var type = this;

            while (true)
            {
                if (!transaction.ChangedParentTypes.TryGetValue(type, out var values))
                {
                    values = new List<YEvent>();
                    transaction.ChangedParentTypes[type] = values;
                }

                values.Add(evt);

                if (type.Item == null)
                {
                    break;
                }

                type = type.Item.Parent as AbstractType;
            }

            InvokeEventHandlers(evt, transaction);
        }

        /// <summary>
        /// Creates YEvent and calls all type observers.
        /// Must be implemented by each type.
        /// </summary>
        public virtual void CallObserver(ITransaction transaction, ISet<string> parentSubs)
        {
            // Do nothing.
        }

        public Item _First()
        {
            var n = Start;
            while (n != null && n.Deleted)
            {
                n = n.Right as Item;
            }
            return n;
        }

        public void InvokeEventHandlers(YEvent evt, ITransaction transaction)
        {
            EventHandler?.Invoke(this, new YEventArgs(evt, transaction));
        }

        public void CallDeepEventHandlerListeners(IList<YEvent> events, ITransaction transaction)
        {
            DeepEventHandler?.Invoke(this, new YDeepEventArgs(events, transaction));
        }

        public string FindRootTypeKey()
        {
            return Doc.FindRootTypeKey(this);
        }

        protected void TypeMapDelete(ITransaction transaction, string key)
        {
            if (Map.TryGetValue(key, out var c))
            {
                c.Delete(transaction);
            }
        }

        protected void TypeMapSet(ITransaction transaction, string key, object value)
        {
            if (!Map.TryGetValue(key, out var left))
            {
                left = null;
            }

            var doc = transaction.Doc;
            var ownClientId = doc.ClientId;
            IContent content;

            if (value == null)
            {
                content = new ContentAny(new object[] { value });
            }
            else
            {
                switch (value)
                {
                    case YDoc d:
                        content = new ContentDoc(d);
                        break;
                    case AbstractType at:
                        content = new ContentType(at);
                        break;
                    case byte[] ba:
                        content = new ContentBinary(ba);
                        break;
                    default:
                        content = new ContentAny(new[] { value });
                        break;
                }
            }

            var newItem = new Item(new StructID(ownClientId, doc.Store.GetState(ownClientId)), left, left?.LastId, null, null, this, key, content);
            newItem.Integrate(transaction, 0);
        }

        protected bool TryTypeMapGet(string key, out object value)
        {
            if (Map.TryGetValue(key, out var val) && !val.Deleted)
            {
                value = val.Content.GetContent()[val.Length - 1];
                return true;
            }

            value = default;
            return false;
        }

        protected object TypeMapGetSnapshot(string key, Snapshot snapshot)
        {
            if (!Map.TryGetValue(key, out var v))
            {
                v = null;
            }

            while (v != null && (!snapshot.StateVector.ContainsKey(v.Id.Client) || v.Id.Clock >= snapshot.StateVector[v.Id.Client]))
            {
                v = v.Left as Item;
            }

            return v != null && v.IsVisible(snapshot) ? v.Content.GetContent()[v.Length - 1] : null;
        }

        protected IEnumerable<KeyValuePair<string, Item>> TypeMapEnumerate() => Map.Where(kvp => !kvp.Value.Deleted);

        protected IEnumerable<KeyValuePair<string, object>> TypeMapEnumerateValues()
        {
            foreach (var kvp in TypeMapEnumerate())
            {
                var key = kvp.Key;
                var value = kvp.Value.Content.GetContent()[kvp.Value.Length - 1];
                yield return new KeyValuePair<string, object>(key, value);
            }
        }
    }
}

File: ycs-golang/Types/AbstractType.go
package types

import (
	"errors"

	"github.com/chenrensong/ygo/structs"
	"github.com/chenrensong/ygo/utils"
)

var (
	ErrNotImplemented = errors.New("not implemented")
)

// YEventArgs represents event arguments for type changes
type YEventArgs struct {
	Event       *utils.YEvent
	Transaction *utils.Transaction
}

// NewYEventArgs creates a new YEventArgs instance
func NewYEventArgs(evt *utils.YEvent, txn *utils.Transaction) *YEventArgs {
	return &YEventArgs{
		Event:       evt,
		Transaction: txn,
	}
}

// YDeepEventArgs represents deep event arguments for nested type changes
type YDeepEventArgs struct {
	Events      []*utils.YEvent
	Transaction *utils.Transaction
}

// NewYDeepEventArgs creates a new YDeepEventArgs instance
func NewYDeepEventArgs(events []*utils.YEvent, txn *utils.Transaction) *YDeepEventArgs {
	return &YDeepEventArgs{
		Events:      events,
		Transaction: txn,
	}
}

// AbstractType is the base type for all Yjs shared types
type AbstractType struct {
	item  *structs.Item
	start *structs.Item
	map_  map[string]*structs.Item

	eventHandlers     []func(*YEventArgs)
	deepEventHandlers []func(*YDeepEventArgs)
}

// NewAbstractType creates a new AbstractType instance
func NewAbstractType() *AbstractType {
	return &AbstractType{
		map_: make(map[string]*structs.Item),
	}
}

// Doc returns the document this type belongs to
func (t *AbstractType) Doc() *utils.YDoc {
	if t.item != nil {
		return t.item.Doc()
	}
	return nil
}

// Parent returns the parent type of this type
func (t *AbstractType) Parent() *AbstractType {
	if t.item != nil {
		if parent, ok := t.item.Parent().(*AbstractType); ok {
			return parent
		}
	}
	return nil
}

// Length returns the length of the type
func (t *AbstractType) Length() int {
	return 0 // To be overridden by concrete types
}

// Integrate integrates this type into the document
func (t *AbstractType) Integrate(doc *utils.YDoc, item *structs.Item) {
	// To be overridden by concrete types
}

// InternalCopy creates an internal copy of the type
func (t *AbstractType) InternalCopy() *AbstractType {
	panic(ErrNotImplemented)
}

// InternalClone creates an internal clone of the type
func (t *AbstractType) InternalClone() *AbstractType {
	panic(ErrNotImplemented)
}

// Write encodes the type to a writer
func (t *AbstractType) Write(encoder utils.Encoder) {
	panic(ErrNotImplemented)
}

// CallTypeObservers notifies observers of type changes
func (t *AbstractType) CallTypeObservers(txn *utils.Transaction, evt *utils.YEvent) {
	current := t
	for current != nil {
		if _, exists := txn.ChangedParentTypes[current]; !exists {
			txn.ChangedParentTypes[current] = []*utils.YEvent{}
		}
		txn.ChangedParentTypes[current] = append(txn.ChangedParentTypes[current], evt)

		if current.item == nil {
			break
		}
		current = current.Parent()
	}

	t.InvokeEventHandlers(evt, txn)
}

// CallObserver calls type observers
func (t *AbstractType) CallObserver(txn *utils.Transaction, parentSubs map[string]struct{}) {
	// To be overridden by concrete types
}

// first returns the first non-deleted item
func (t *AbstractType) first() *structs.Item {
	n := t.start
	for n != nil && n.Deleted() {
		n = n.Right()
	}
	return n
}

// InvokeEventHandlers invokes registered event handlers
func (t *AbstractType) InvokeEventHandlers(evt *utils.YEvent, txn *utils.Transaction) {
	args := NewYEventArgs(evt, txn)
	for _, handler := range t.eventHandlers {
		handler(args)
	}
}

// CallDeepEventHandlerListeners invokes deep event handlers
func (t *AbstractType) CallDeepEventHandlerListeners(events []*utils.YEvent, txn *utils.Transaction) {
	args := NewYDeepEventArgs(events, txn)
	for _, handler := range t.deepEventHandlers {
		handler(args)
	}
}

// FindRootTypeKey finds the root type key
func (t *AbstractType) FindRootTypeKey() string {
	if t.Doc() != nil {
		return t.Doc().FindRootTypeKey(t)
	}
	return ""
}

// TypeMapDelete deletes a value from the type map
func (t *AbstractType) TypeMapDelete(txn *utils.Transaction, key string) {
	if item, exists := t.map_[key]; exists {
		item.Delete(txn)
	}
}

// TypeMapSet sets a value in the type map
func (t *AbstractType) TypeMapSet(txn *utils.Transaction, key string, value interface{}) {
	var left *structs.Item
	if existing, exists := t.map_[key]; exists {
		left = existing
	}

	doc := txn.Doc
	ownClientID := doc.ClientID()
	var content structs.Content

	switch v := value.(type) {
	case *utils.YDoc:
		content = structs.NewContentDoc(v)
	case *AbstractType:
		content = structs.NewContentType(v)
	case []byte:
		content = structs.NewContentBinary(v)
	case nil:
		content = structs.NewContentAny([]interface{}{nil})
	default:
		content = structs.NewContentAny([]interface{}{value})
	}

	newItem := structs.NewItem(
		utils.NewID(ownClientID, doc.Store.GetState(ownClientID)),
		left,
		left.LastID(),
		nil,
		nil,
		t,
		key,
		content,
	)
	newItem.Integrate(txn, 0)
}

// TryTypeMapGet attempts to get a value from the type map
func (t *AbstractType) TryTypeMapGet(key string) (interface{}, bool) {
	if item, exists := t.map_[key]; exists && !item.Deleted() {
		content := item.Content().GetContent()
		if len(content) > 0 {
			return content[len(content)-1], true
		}
	}
	return nil, false
}

// TypeMapGetSnapshot gets a value from the type map at a specific snapshot
func (t *AbstractType) TypeMapGetSnapshot(key string, snapshot *utils.Snapshot) interface{} {
	item, exists := t.map_[key]
	if !exists {
		item = nil
	}

	for item != nil {
		clientState, hasClient := snapshot.StateVector[item.ID().Client]
		if !hasClient || item.ID().Clock >= clientState {
			item = item.Left()
			continue
		}

		if item.IsVisible(snapshot) {
			content := item.Content().GetContent()
			if len(content) > 0 {
				return content[len(content)-1]
			}
		}
		break
	}

	return nil
}

// TypeMapEnumerate enumerates all items in the type map
func (t *AbstractType) TypeMapEnumerate() map[string]*structs.Item {
	result := make(map[string]*structs.Item)
	for k, v := range t.map_ {
		if !v.Deleted() {
			result[k] = v
		}
	}
	return result
}

// TypeMapEnumerateValues enumerates all values in the type map
func (t *AbstractType) TypeMapEnumerateValues() map[string]interface{} {
	result := make(map[string]interface{})
	for k, v := range t.TypeMapEnumerate() {
		content := v.Content().GetContent()
		if len(content) > 0 {
			result[k] = content[len(content)-1]
		}
	}
	return result
}
